File: mvpcore/src/main/java/biz/dealnote/mvp/compat/SimplePresenterLoader.java
package biz.dealnote.mvp.compat;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.Loader;

import biz.dealnote.mvp.core.IMvpView;
import biz.dealnote.mvp.core.IPresenter;
import biz.dealnote.mvp.core.IPresenterFactory;

public class SimplePresenterLoader<P extends IPresenter<V>, V extends IMvpView> extends Loader<P> {

    private IPresenterFactory<P> factory;

    SimplePresenterLoader(@NonNull Context context, @NonNull IPresenterFactory<P> factory) {
        super(context);
        this.factory = factory;
    }

    private P presenter;

    @NonNull
    public P get(){
        if(presenter == null){
            presenter = factory.create();
            factory = null;
        }

        return presenter;
    }

    @Override
    protected void onReset() {
        super.onReset();
        if (presenter != null) {
            presenter.destroy();
            presenter = null;
        }
    }
}

File: mvpcore/src/main/java/biz/dealnote/mvp/compat/SimplePresenterLoader.kt
package biz.dealnote.mvp.compat

import android.content.Context
import android.support.v4.content.Loader

import biz.dealnote.mvp.core.IMvpView
import biz.dealnote.mvp.core.IPresenter
import biz.dealnote.mvp.core.IPresenterFactory

class SimplePresenterLoader<P : IPresenter<V>, V : IMvpView> constructor(context: Context, var factory: IPresenterFactory<P>) : Loader<P>(context) {

    private var f: IPresenterFactory<P>? = factory

    private var presenter: P? = null

    fun get(): P {
        if (presenter == null) {
            presenter = factory.create()
            f = null
        }

        return presenter!!
    }

    override fun onReset() {
        super.onReset()
        presenter?.destroy()
        presenter = null
    }
}